src/google/protobuf/compiler/csharp/csharp_helpers.h

```diff
@@ -158,6 +158,34 @@ inline bool IsProto2(const FileDescriptor* descriptor) {
   return descriptor->syntax() == FileDescriptor::SYNTAX_PROTO2;
 }
 
+inline bool SupportsPresenceApi(const FieldDescriptor* descriptor) {
+  // Not quite the same as descriptor->is_singular_with_presence() in two ways:
+  
+  // In proto3 we add field presence only for explicitly-optional fields. Message fields
+  // don't get field presence by default, because the user can just compare with a null reference.
+  
+  // We generate Has/Clear for oneof fields in C#.
+  // It's possible that we shouldn't, but stopping doing so would be a breaking change for proto2.
```

Comment:
I think it's good to have has (at least) for oneof fields. I'm going to propose that we do this across the board.

Reply:
I would like to keep this initial PR as *just* introducing optional fields for proto3 - deliberately not changing any other code, beyond the previously-inefficient presence bit allocation (which would be more significant in proto3 than in proto2). I'm happy to discuss other API changes and make those in a separate PR, but I'll be a lot more confident in *this* PR if existing proto3 code doesn't change at all.

Reply:
That's totally fine, I just wanted to free you (and this comment) from the feeling of obligation that you "should" be suppressing has/clear for oneof fields.

When I said "I'm going to propose" I meant I was going to make this proposal separately, as a statement of intent for what all languages should aim for. I'm not proposing it for this review. :)

Reply:
SGTM :)

Reply:
Done.